File: fake_nlp.py
import re
from fuzzywuzzy import fuzz
import nltk
from itertools import combinations as combos
from copy import deepcopy

# Download NLTK resources (Only needs to be run once)
nltk.download('stopwords')

# Expanded stopwords list
expanded_stopwords = [
    'hello', 'hi', 'hey', 'greetings', 'morning', 'evening', 'okay', 'thanks', 
    'please', 'sorry', 'sure', 'wow', 'oops', 'ah', 'uh', 'well', 'I', 'you', 
    'me', 'your', 'my', 'the', 'a', 'an', 'is', 'it', 'to', 'of', 'and', 'for', 
    'on', 'at', 'by', 'with', 'about', 'as', 'from', 'but', 'not', 'this', 
    'that', 'which', 'who', 'whom', 'their', 'ours', 'ourselves', 'they', 
    'them', 'we', 'us', 'ourselves', 'all', 'any', 'how', 'how’s', 'what', 'what’s', 
    'where', 'where’s', 'when', 'when’s', 'why', 'why’s', 'can', 'could', 'will', 
    'would', 'shall', 'should', 'may', 'might', 'must', 'ought', 'each', 'every', 
    'anybody', 'someone', 'thing', 'something','got','get'
]

# Combine with NLTK stopwords
stopwords = set(nltk.corpus.stopwords.words('english')) | set(expanded_stopwords)

# List of symptoms directly in code
symptoms_list = []
with open('try.txt','r') as f:
    symptoms_list.extend(f.read().split('\n'))

# Beautify the input text (remove stopwords, punctuation)
def beautify_string(text):
    text = text.lower()  # Convert to lowercase
    text = re.sub(r'\W', ' ', text)  # Remove punctuation
    text = re.sub(r'\s+', ' ', text)  # Replace multiple spaces with a single space
    words = text.split()  # Tokenize into words
    words = [word for word in words if word not in stopwords]  # Remove stopwords
    return words

# Function to check for symptoms in the user input
def check_symptoms_optimization(user_input, symptoms_list, threshold=55):
    user_input_words = beautify_string(user_input)
    found_symptoms = []

    # Check fuzzy matching with symptoms list
    for symptom in symptoms_list:
        if fuzz.partial_ratio(' '.join(user_input_words), symptom.lower()) > threshold:
            found_symptoms.append(symptom)
    return found_symptoms, user_input_words

def check_symptoms(user_input,symptoms_list,threshold=55):
    symptoms_list,user_input = check_symptoms_optimization(user_input,symptoms_list,30)
    ansset = set()
    inp_dict =  dict()
    for i in user_input:
        inp_dict[i] = set(check_symptoms_optimization(i,symptoms_list,80)[0])

    keys_dict = list(inp_dict.keys())
    for i in keys_dict:
        if len(inp_dict[i]) == 0:
            del inp_dict[i]
    keys_dict = list(inp_dict.keys())
    i = 5
    while i != 0:
        combinations = combos(keys_dict,i)
        for combo in combinations:
            newset = deepcopy(inp_dict[combo[0]])
            for y in range(i):
                if y == 0:
                    continue
                newset.intersection_update(inp_dict[combo[y]])
            if len(newset):
                ansset = ansset|newset
    
        i -= 1

    tempset = deepcopy(ansset)
    for symptoms in tempset:
        le_count = 0
        for keywords in keys_dict:
            for eachword in symptoms.split():
                # print(keywords,eachword)
                if re.match(keywords,eachword) != None:
                    le_count += 1
                    break
        if (le_count/len(symptoms.split())) < threshold/100:
            ansset.remove(symptoms)

    if len(ansset) == 0:
        return [' '.join(user_input)]
    return ansset


# Main chatbot loop
def chatbot():
    print("Welcome to the Health Chatbot!")
    print("Tell me your symptoms or type 'exit' to quit.")
    
    while True:
        # Take user input
        user_input = input("You: ")
        
        if user_input.lower() == 'exit':
            print("Goodbye!")
            break
        
        # Check for symptoms
        check_symptoms(user_input, symptoms_list)

if __name__ == '__main__':
    # Run the chatbot
    chatbot()

File: fakepa-server/web/fakenlp.js
function removeStopwords(inputList, stopwords) {
    // Filter out the words in inputList that are present in stopwords
    return inputList.filter(word => !stopwords.includes(word));
}

function tokenizeString(inputString) {
    // Remove underscores, then match words and convert to lowercase
    const tokens = inputString
        .replace(/_/g, ' ') // Replace underscores with spaces
        .match(/\b[a-zA-Z0-9']+\b/g); // Match words containing letters, digits, and apostrophes

    return tokens ? tokens.map(token => token.toLowerCase()) : [];
}

// Function to compute the dot product of two vectors
function dotProduct(A, B) {

    return A.reduce((sum, a, index) => sum + a * B[index], 0);
  }
  
  // Function to compute the magnitude (or norm) of a vector
function magnitude(A) {
    return Math.sqrt(A.reduce((sum, a) => sum + a * a, 0));
  }
  
  // Function to compute the cosine similarity between two vectors
function cosineSimilarity(A, B) {
    const dot = dotProduct(A, B);
    const magA = magnitude(A);
    const magB = magnitude(B);
    
    // Check to prevent division by zero
    if (magA === 0 || magB === 0) {
      return 0; // If either vector has magnitude 0, return similarity as 0
    }
  
    return dot / (magA * magB);
  }

  // Function to sort an array of vectors based on the cosine similarity to the given vector, with a threshold
function sortVectorsByCosineSimilarity(vectors, targetVector, threshold = 0) {
    // Create an array of tuples: [cosine similarity, index]
    const similarities = vectors.map((vector, index) => {
      const similarity = cosineSimilarity(vector, targetVector);
      return {
        index: index,          // Store the original index
        similarity: similarity // Store the similarity value
      };
    });
  
    // Log similarities for debugging
    console.log('Similarities with Indices:', similarities);
  
    // Filter out vectors with similarity less than the threshold
    const filteredSimilarities = similarities.filter(item => item.similarity >= threshold);
  
    // Log filtered similarities
    console.log('Filtered similarities:', filteredSimilarities);
  
    // Sort the similarities by cosine similarity in descending order
    filteredSimilarities.sort((a, b) => b.similarity - a.similarity);
  
    // Return the indices of the vectors sorted by similarity
    return filteredSimilarities.map(item => item.index);
  }

var vocab = (await (await fetch('/vocab.txt')).text()).split(/\s/); //can also be .json() or .other_set_type
var vectors = await (await fetch('/vectors.txt')).text();
vectors = vectors.split('\n');
var words = (await (await fetch('/words.txt')).text()).split('\n');

const str = "I have a ache in stomach" ;
const tokens = tokenizeString(str);
const stopwords = ["youve", 'have', 'only', "isnt", 'll', 'may', "hadnt", 'their', 'about', "theyll", 'will', 'well', "dont", 'there', 'ain', 'ours', 'ought', 'so', 'having', 'whom', 'all', 'ourselves', 'am', 'what', "mustnt", 'o', 'needn', 'because', "theyd", 'been', 'who', "theyve", 'until', "arent", 'some', 'in', 'i', 'nor', 'other', 'must', 'wheres', 'for', 'them', 'every', 'uh', 'any', 'its', 'evening', 'anybody', 'themselves', 'down', 'hadn', 'd', "doesnt", 'but', "thatll", 'itself', "weve", 'a', 'before', "didnt", 'further', 'under', "neednt", 'isn', "shes", 'you', 'weren', 'won', "ill", 'has', "shed", 'here', 'someone', 'don', 'why', 'how', 'own', 'oops', "couldnt", 'shan', 'sorry', 'during', 'over', 'into', 'most', 'when', 'our', 'these', 'more', 'to', 'with', 'thing', 'of', 'as', 'does', 'can', 'is', 'get', 'his', 'he', "wasnt", 'which', 'your', 'didn', 'hey', 'at', 'ah', 'out', 'mustn', 'hers', 'from', "hes", 'aren', 'I', 'himself', 'mightn', 've', 'both', 'yourself', "youre", 'after', 'same', 'off', 'm', 're', "well", 'this', 'could', "id", "im", 'through', 'wouldn', "youd", 'okay', 'got', 'yours', 'would', 'greetings', 'shouldn', "mightnt", 'then', 'too', 'an', 'below', 'whens', 'wasn', 'my', 'shall', 'hasn', 'him', "hasnt", 'were', 'be', 'doesn', "were", "ive", 'they', 'thanks', 'each', "werent", 'and', "havent", 'against', 'between', 'she', 'the', 'haven', "shell", 'are', "itd", 'whys', 'wow', 'whats', 'hows', "shouldnt", 'might', 'yourselves', 'do', 'again', 'me', "shouldve", 'had', 'very', 'should', 'if', 't', 'hi', "youll", 'being', 'y', "shant", 'myself', 'up', "wont", 'no', 'was', 'hello', 'by', 'such', 'that', 'her', 'than', 'once', "its", 'herself', 'doing', "hed", 'couldn', 'not', 'we', "wed", 'above', 'now', 'few', 'while', 'please', 'on', 'did', 'something', "itll", 'sure', 'those', "hell", 's', 'it', 'morning', 'us', 'theirs', "theyre", 'where', "wouldnt", 'just', 'or', 'ma'];
//console.log(tokens);
const worklist = removeStopwords(tokens, stopwords);
//console.log(worklist);

let matrix = [];

for (let i = 0; i < vectors.length; i++) {
    matrix[i] = Array.from(vectors[i], char => Number(char));    
}
let sentenceVec = vocab.map(word => {
    // Count occurrences of the word in smallerArr if it exists in arr
    return worklist.filter(item => item === word).length;
  });

const threshold = 0.5;
const sortedIndices = sortVectorsByCosineSimilarity(matrix, sentenceVec, threshold);
for (let i = 0; i < sortedIndices.length; i++){
    console.log(words[sortedIndices[i]]);
}